src/ui/reducers/timeline.ts

```diff
@@ -42,6 +45,12 @@ const timelineSlice = createSlice({
     setTrimRegion(state, action: PayloadAction<FocusRegion | null>) {
       state.focusRegion = action.payload;
     },
+    pointsReceived(state, action: PayloadAction<TimeStampedPoint[]>) {
+      state.points = sortBy(
+        uniqBy(state.points.concat(action.payload), p => p.time),
+        p => p.time
+      );
+    },
```

Comment:
This code is succinct but it also looks kind of expensive. How often do we think it will run?

For example, I think we could write a more performant version like this (but admittedly it's a lot less readable):
```js
const points = state.points.concat();

let actionIndex = 0;
let stateIndex = 0;

while (actionIndex < action.payload.length) {
  const actionPoint = action.payload[actionIndex];

  let stateIndexPoint = points[stateIndex];

  if (stateIndexPoint == null) {
    points.push(actionPoint);
    actionIndex++;
  } else if (actionPoint.time === stateIndexPoint.time) {
    // Don't add duplicates.
    actionIndex++;
  } else if (actionPoint.time < stateIndexPoint.time) {
    points.splice(stateIndex, 0, actionPoint);
    actionIndex++;
    stateIndex++;
  } else {
    stateIndex++;
  }
}
```

I think this would be significantly faster (see [jsben.ch/xtKWW](https://jsben.ch/xtKWW)) but maybe it's not worth unless we think it's likely to be run a lot?

In general I'm concerned about how many inline maps and sorts we do– on particularly large sets of data– without memoization.

Reply:
yeah, I was definitely a little lazy when writing this, I'm happy to add a more tuned version.